src/plots/ternary/ternary.js

```diff
@@ -0,0 +1,325 @@
+/**
+* Copyright 2012-2016, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var d3 = require('d3');
+var Color = require('../../components/color');
+var Drawing = require('../../components/drawing');
+var setConvert = require('../cartesian/set_convert');
+var extendFlat = require('../../lib/extend').extendFlat;
+var Axes = require('../cartesian/axes');
+var filterVisible = require('../../lib/filter_visible');
+
+
+function Ternary(options, fullLayout) {
+    this.id = options.id;
+    this.graphDiv = options.graphDiv;
+    this.container = fullLayout._ternarylayer;
+    this.defs = fullLayout._defs;
+    this.layoutId = fullLayout._uid;
+    this.traceHash = {};
+
+    this.makeFramework();
+}
+
+module.exports = Ternary;
+
+var proto = Ternary.prototype;
+
+proto.plot = function(ternaryData, fullLayout) {
+    var _this = this,
+        ternaryLayout = fullLayout[_this.id],
+        graphSize = fullLayout._size,
+        i;
+
+    _this.adjustLayout(ternaryLayout, graphSize);
+
+    var traceHashOld = _this.traceHash;
+    var traceHash = {};
+
+    for(i = 0; i < ternaryData.length; i++) {
+        var trace = ternaryData[i];
+
+        traceHash[trace.type] = traceHash[trace.type] || [];
+        traceHash[trace.type].push(trace);
+    }
+
+    var moduleNamesOld = Object.keys(traceHashOld);
+    var moduleNames = Object.keys(traceHash);
+
+    // when a trace gets deleted, make sure that its module's
+    // plot method is called so that it is properly
+    // removed from the DOM.
+    for(i = 0; i < moduleNamesOld.length; i++) {
+        var moduleName = moduleNamesOld[i];
+
+        if(moduleNames.indexOf(moduleName) === -1) {
+            var fakeModule = traceHashOld[moduleName][0];
+            fakeModule.visible = false;
+            traceHash[moduleName] = [fakeModule];
+        }
+    }
```

Comment:
This part :arrow_double_up: isn't great. But, thanks for making `ternary` as good (or as bad depending how you look at it) as `geo`. 

One day, I hope to remove the per-trace-type grouping making the `traceHash` obsolete. This implies that the plot step of all trace modules would be free of cross-trace computations (e.g. `scatter` lines with fill need have cross-trace plot step). I'm thinking all cross-trace computations should be moved to `setPositions`. But, that might harder than it seems.